desc/compute/utils.py

```diff
@@ -649,63 +781,120 @@ def surface_integrals(grid, q=jnp.array([1]), surface_label="rho", max_surface=F
                 "yellow",
             )
         )
-
-    q = jnp.atleast_1d(q)
-    nodes, unique_idx, _, ds, max_surface_val, has_endpoint_dupe = _get_grid_surface(
+    unique_size, inverse_idx, spacing, has_endpoint_dupe = _get_grid_surface(
         grid, surface_label
     )
 
-    if max_surface:
-        # assumes surface label is zeta, does a line integral
-        max_rho = grid.nodes[grid.unique_rho_idx[-1], 0]
-        idx = np.nonzero(grid.nodes[:, 0] == max_rho)[0]
-        q = q[idx]
-        nodes = nodes[idx]
-        unique_idx = (unique_idx / grid.num_rho).astype(int)
-        ds = ds[idx] / grid.spacing[idx, 0] / max_rho
-
-    # Separate nodes into bins with boundaries at unique values of the surface label.
-    # This groups nodes with identical surface label values.
-    # Each is assigned a weight of their contribution to the integral.
-    # The elements of each bin are summed, performing the integration.
-    bins = jnp.append(nodes[unique_idx], max_surface_val)
-
-    # to group duplicate nodes together
-    nodes_modulo = nodes % max_surface_val if has_endpoint_dupe else nodes
-    # Longer explanation:
-    #     Imagine a torus cross-section at zeta=pi.
-    # A grid with a duplicate zeta=pi node has 2 of those cross-sections.
-    #     In grid.py, we multiply by 1/n the areas of surfaces with
-    # duplicity n. This prevents the area of that surface from being
-    # double-counted, as surfaces with the same node value are combined
-    # into 1 integral, which sums their areas. Thus, if the zeta=pi
-    # cross-section has duplicity 2, we ensure that the area on the zeta=pi
-    # surface will have the correct total area of pi+pi = 2pi.
-    #     An edge case exists if the duplicate surface has nodes with
-    # different values for the surface label, which only occurs when
-    # has_endpoint_dupe is true. If has_endpoint_dupe is true, this grid
-    # has a duplicate surface at surface_label=0 and
-    # surface_label=max_surface_val. Although the modulo of these values
-    # are equal, their numeric values are not, so jnp.histogram would treat
-    # them as different surfaces. We solve this issue by modulating
-    # 'nodes', so that the duplicate surface has nodes with the same
-    # surface label and is treated as one, like in the previous paragraph.
-
-    integrals = jnp.histogram(nodes_modulo, bins=bins, weights=ds * q)[0]
+    # Todo: Define masks as a sparse matrix once sparse matrices are
+    #       are no longer experimental in jax.
+    # The ith row of masks is True only at the indices which correspond to the
+    # ith surface. The integral over the ith surface is the dot product of the
+    # ith row vector and the vector of integrands of all surfaces.
+    masks = inverse_idx == jnp.arange(unique_size)[:, jnp.newaxis]
```

Comment:
This is a boolean array, so even though it's shape is N x N^3, its memory is N^4/8 <= 3N^3 = grid.nodes almost always, and therefore much smaller than the grid itself.

Reply:
is this $N^{4/8}$ or $N^4 / 8$?

Reply:
the latter. seems scary but it actually uses less memory (https://github.com/PlasmaControl/DESC/pull/501#issuecomment-1546307972) than the implementation on master. I think the idea is that matrix operations are easy to differentiate through.

Reply:
The sparse matrix is actually implementable now with `masks = sparse.BCOO.fromdense(inverse_idx == jnp.arange(unique_size)[:, jnp.newaxis])`. but it didn't seem worth it